### PR TITLE
PlatformViewLink handles focus

### PR DESCRIPTION
**Pull request by cyanglaz**
## Description

* In the `build` of `PlatformViewLink`, Added a FocusNode wrapping around the surface widget.
* The focus node will ask platform view to clear its focus when necessary through [PlatformViewController.clearFocus].
* The platform view can notify the framework it wants to gain focus by calling [PlatformViewCreationParams.onFocusChanged]

## Related Issues

https://github.com/flutter/flutter/issues/36779

## Tests

I added the following tests:

widgets/platform_view_test.dart
* PlatformViewLink manages the focus properly

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 1]
```diff
@@ -2041,5 +2041,67 @@ void main() {
...
+    testWidgets('PlatformViewLink manages the focus properly', (WidgetTester tester) async {
+      final GlobalKey containerKey = GlobalKey();
+      FakePlatformViewController controller;
+      ValueChanged<bool> focusChanged;
+      final PlatformViewLink platformViewLink = PlatformViewLink(createPlatformViewController: (PlatformViewCreationParams params){
+        params.onPlatformViewCreated(params.id);
+        focusChanged = params.onFocusChanged;
+        controller = FakePlatformViewController(params.id);
+        return controller;
+      }, surfaceFactory: (BuildContext context, PlatformViewController controller) {
```
Comment (amirh) [inline packages/flutter/test/widgets/platform_view_test.dart]: nit: start surfaceFactory in a new line

File: packages/flutter/test/widgets/platform_view_test.dart [hunk 2]
```diff
@@ -2041,5 +2041,67 @@ void main() {
...
+
+      expect(containerFocusNode.hasFocus, true);
+      expect(platformViewFocusNode.hasFocus, false);
+
+      // ask the platform view to gain focus
+      focusChanged(true);
+      await tester.pump();
+
+      expect(containerFocusNode.hasFocus, false);
+      expect(platformViewFocusNode.hasFocus, true);
```
Comment (amirh) [inline packages/flutter/test/widgets/platform_view_test.dart]: Assert that `controller.focusCleared` is false.

File: packages/flutter/lib/src/services/platform_views.dart [hunk 3]
```diff
@@ -734,4 +734,7 @@ abstract class PlatformViewController {
   ///
   /// The [PlatformViewController] is unusable after calling dispose.
   void dispose();
+
+  /// Clears the focus of the platform view.
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter/lib/src/services/platform_views.dart]: done

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 4]
```diff
@@ -599,6 +600,9 @@ class PlatformViewCreationParams {
 
   /// Callback invoked after the platform view has been created.
   final PlatformViewCreatedCallback onPlatformViewCreated;
+
+  /// Notifies the [PlatformViewLink] when the focus is changed from the platform.
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 5]
```diff
@@ -679,31 +683,55 @@ class _PlatformViewLinkState extends State<PlatformViewLink> {
...
+        onFocusChanged:_onPlatformFocusChange
+      ),
+    );
   }
 
   void _onPlatformViewCreated(int id) {
     setState(() => _platformViewCreated = true);
   }
 
+  void _onFrameworkFocusChange(bool isFocused) {
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done

**Review by cyanglaz — COMMENTED** [review]
Updated with review comments.

File: packages/flutter/lib/src/widgets/platform_view.dart [hunk 6]
```diff
@@ -679,31 +685,55 @@ class _PlatformViewLinkState extends State<PlatformViewLink> {
...
+    );
   }
 
   void _onPlatformViewCreated(int id) {
     setState(() => _platformViewCreated = true);
   }
 
+  void _handleFrameworkFocusChanged(bool isFocused) {
+    if (!isFocused) {
+      _controller?.clearFocus();
```
Comment (amirh) [inline packages/flutter/lib/src/widgets/platform_view.dart]: Let's set `_controller` to null after calling dispose, to make sure we're not accidentally doing this on a disposed controller.

**Review by amirh — APPROVED** [review]
LGTM (left one comment)

File: packages/flutter/lib/src/widgets/platform_view.dart (hunk 6 above)
Comment (cyanglaz) [reply packages/flutter/lib/src/widgets/platform_view.dart]: done! Going to merge on green. Thanks for the quick turnaround.

**Comment by goderbauer** [comment]
@cyanglaz Can you update the PR description to better describe what is actually contained in this change. What functionality is added? What is changed? And why? This will help future archeologists looking through the history.

**Comment by cyanglaz** [comment]
> @cyanglaz Can you update the PR description to better describe what is actually contained in this change. What functionality is added? What is changed? And why? This will help future archeologists looking through the history.

Updated the description! Let me know if it makes sense now. :)

**Pull request merged**